Web/www/index.html

```diff
@@ -0,0 +1,194 @@
+					// Labels
+					var label_prop = document.createElement("label");
+					label_prop.style = "display:inline-block; width:10%;";
+					label_prop.class = "text--medium";
+					var label_integ = label_prop.cloneNode(true);
+					var label_deriv = label_prop.cloneNode(true);
+					label_prop.innerHTML = "P";
+					label_integ.innerHTML = "I";
+					label_deriv.innerHTML = "D";
+					// Inputs
+					var prop = document.createElement("input");
+					prop.type = "text";
+					prop.class = "card_input";
+					var integ = prop.cloneNode(true);
+					var deriv = prop.cloneNode(true);
+					prop.id = names[i] + "_p";
+					integ.id = names[i] + "_i";
+					deriv.id = names[i] + "_d";
+					prop.value = data[i].p;
+					integ.value = data[i].i;
+					deriv.value = data[i].d;
+					// Increments
+					var prop_up = document.createElement("button");
+					prop_up.type = "increment";
+					prop_up.className = "button_up text--bold";
+					prop_up.innerHTML = "∧";
+					var prop_down = prop_up.cloneNode(true);
+					prop_down.className = "button_down text--bold";
+					prop_down.innerHTML = "∨";
+					var integ_up = prop_up.cloneNode(true);
+					var integ_down = prop_down.cloneNode(true);
+					var deriv_up = prop_up.cloneNode(true);
+					var deriv_down = prop_down.cloneNode(true);
+					prop_up.id = names[i] + "_but_p_up";
+					prop_down.id = names[i] + "_but_p_d";
+					integ_up.id = names[i] + "_but_i_up";
+					integ_down.id = names[i] + "_but_i_d";
+					deriv_up.id = names[i] + "_but_d_up";
+					deriv_down.id = names[i] + "_but_d_d";
+
+					// Assign values to inputs
+					pid_card.appendChild(_br.cloneNode(true));
+					pid_card.appendChild(title);
+					pid_card.appendChild(_br.cloneNode(true));
+
+					pid_card.appendChild(label_prop);
+					pid_card.appendChild(prop);
+					pid_card.appendChild(prop_up.cloneNode(true));
+					pid_card.appendChild(prop_down.cloneNode(true));
+					pid_card.appendChild(_br.cloneNode(true));
+
+					pid_card.appendChild(label_integ);
+					pid_card.appendChild(integ);
+					pid_card.appendChild(integ_up.cloneNode(true));
+					pid_card.appendChild(integ_down.cloneNode(true));
+					pid_card.appendChild(_br.cloneNode(true));
+
+					pid_card.appendChild(label_deriv);
+					pid_card.appendChild(deriv);
+					pid_card.appendChild(deriv_up.cloneNode(true));
+					pid_card.appendChild(deriv_down.cloneNode(true));
+
+					pid_card.appendChild(_br.cloneNode(true));
+				}
+				var submit = document.createElement("button");
+				submit.className = "text--bold";
+				submit.type = "submit";
+				submit.innerHTML = "Submit";
+				submit.id = "submit_btn";
+				var submit_contain = document.createElement("div");
+				submit_contain.style = "text-align: center;";
+				submit_contain.appendChild(submit);
+				pid_card.appendChild(_br.cloneNode(true));
+				pid_card.appendChild(submit_contain);
+			}
+			window.onload = function(){
+				// Submit button listener
+				// Should read all of the text boxes and save to pid.json
```

Comment:
This is probably quite confusing alright. I haven't implemented the part to read and use the PID parameters anywhere else, but this is the thinking behind it...

I want to keep the PID values in non-volatile memory so that they are maintained through power cycles, but I also want programs to be able to modify them. Traditionally in embedded stuff, the program sits in flash and any non-volatile variables sit on an EEPROM (generally the flash is not used at runtime, just to store the program). Since the ESP has no EEPROM, a virtual EEPROM can be made if the arduino IDE is being used to program the device (using EEPROM.h).

However, the ESP32 is running micropython :) Therefore the microcontroller has a filesystem built in which allows flash memory to be accessed during runtime (which is pretty cool for a microcontroller). The server runs asynchronously and reads/writes to the JSON file (pid.json) on the ESP. The main glider script that runs the PID controllers can also access this file and update the PID parameters based on the values inside.

I think that some pyhtml would work well actually! Maybe I could write directly to JSON file?

TLDR; The JSON file is acting as a non-volatile area to store the PID parameters. The file doesn't come up anywhere else because I haven't implemented the back-end stuff yet.

Reply:
So just for a more high level understanding, when does the ESP 32 read `pid.json`? Just whenever it boots up?

Reply:
It will read from it on a regular basis, so about every second or so. That way you can change the PID parameters in roughly real time and see how the plane reacts.

Reply:
Ah cool cool yeah that makes sense. Do you have the name of the HTML template generator library that uses the `.pyhtml` files by chance actually? Can't seem to find it online and it looks pretty cool.

Reply:
That's why I didn't look too far into it. However, if you go to https://github.com/jczic/MicroWebSrv and then open microWebTemplate.py, that should give you an idea. MicroWebSrv is the server that runs on the ESP32 in micropython 😄